storjscandb/wallets.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2022 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storjscandb
+
+import (
+	"context"
+	"time"
+
+	"github.com/zeebo/errs"
+	"storj.io/storjscan/storjscandb/dbx"
+)
+
+// ErrWalletsDB indicates about internal wallets DB error.
+var ErrWalletsDB = errs.Class("WalletsDB")
+
+// WalletsDB is wallets database dbx postgres implementation that stores deposit address information.
+//
+// architecture: Database
+type WalletsDB struct {
+	db *dbx.DB
+}
+
+// Wallet represents an entry in the wallets table.
+type Wallet struct {
+	Address []byte
+	Claimed time.Time
+}
+
+// Create inserts a new entry in the wallets table.
+func (wallets *WalletsDB) Create(ctx context.Context, address []byte) (Wallet, error) {
+	w, err := wallets.db.Create_Wallet(ctx, dbx.Wallet_Address(address), dbx.Wallet_Create_Fields{})
+	return Wallet{Address: w.Address}, ErrWalletsDB.Wrap(err)
+}
+
+// CreateBatch inserts a new db entry for each address.
+func (wallets *WalletsDB) CreateBatch(ctx context.Context, addresses [][]byte) (err error) {
+	err = wallets.db.WithTx(ctx, func(ctx context.Context, tx *dbx.Tx) error {
+		for _, address := range addresses {
+			_, err = tx.Create_Wallet(ctx, dbx.Wallet_Address(address), dbx.Wallet_Create_Fields{})
+			if err != nil {
+				return err
+			}
+		}
+		return err
+	})
+	return ErrWalletsDB.Wrap(err)
+}
+
+// Get queries the wallets table for the information stored for a given address
+func (wallets *WalletsDB) Get(ctx context.Context, address []byte) (Wallet, error) {
+	w, err := wallets.db.Get_Wallet_By_Address(ctx, dbx.Wallet_Address(address))
+	return Wallet{Address: w.Address, Claimed: *w.Claimed}, ErrWalletsDB.Wrap(err)
+}
+
+// GetNextAvailable returns the first unclaimed wallet address.
+func (wallets *WalletsDB) GetNextAvailable(ctx context.Context) (Wallet, error) {
+	w, err := wallets.db.First_Wallet_By_Claimed_Is_Null(ctx)
+	return Wallet{Address: w.Address}, ErrWalletsDB.Wrap(err)
+}
+
+// Claim sets the timestamp at which a wallet address is claimed.
+func (wallets *WalletsDB) Claim(ctx context.Context, address []byte) (Wallet, error) {
```

Comment:
Is there ever a case in which the need may arise to `unclaim` a wallet?

Reply:
As an MVP we probably aren't going to allow users to unclaim a wallet, but perhaps I should just create a method to do so in case we want to in the future?

Reply:
I guess, the use case I was thinking about is if I start paying for my projects in storj token, but then I cancel/delete my account. What happens to the wallet that was generated for me?

Reply:
I suppose there could be a couple of options. Unsure which would be better for historical accounting purposes or if we care.
1. The wallet is still marked as claimed and no one uses it.
2. We can unlink the wallet and the account and then mark it as unclaimed.

Reply:
resolution: allow addresses to remain claimed